pandas/core/describe.py

```diff
@@ -91,7 +91,7 @@ def describe_ndframe(
         )
 
     result = describer.describe(percentiles=percentiles)
-    return cast(FrameOrSeries, result)
+    return cast(NDFrameT, result)
```

Comment:
The body of `describe_ndframe` makes it clear that it does not work with any NDFrame: Series-like is treated as a Series and DataFrame-like is treated as a DataFrame.

The TypeVar is "correct" - replacing print(type()) in your first example with reveal_type:
bar.py:3: note: Revealed type is "pandas.core.frame.DataFrame*"
bar.py:5: note: Revealed type is "pandas.core.series.Series*"
bar.py:12: note: Revealed type is "bar.DataFrameSubClass*"
bar.py:19: note: Revealed type is "bar.SeriesSubClass*"

I think that `NDFrameT` is probably still the best annotation here.

Reply:
`FrameOrSeriesT` would be better, until describe is generic enough (if that is a goal)

Reply:
It wouldn't make much of a difference: TypeVar is covariant by default, it would suggest that sub-classes are also handled correctly (which they are not in the case of describe). Could make an invariant TypeVar or leave as-is?